### ISSUE TITLE
FR: Hide *.sdr Folders

### ISSUE THREAD

**Issue by liberodark** [koreader/koreader#8577]
Hi,

I use your software on PocketBook PB740 & PB741
The main issue for me is the creation of `Name.sdr` on the all folders when you read a ebook.
The problem is that if you define the Koreader as the default reader.
And that we navigate with the file manager of PocketBook.
It displays all `Name.sdr` folders.
The solution that seems the easiest to no longer see them is simply to put them in `.Name.sdr` which has the effect of putting them in a hidden folder and therefore not visible in the explorer.
Unfortunately lua is not my forte.
I have looked at the code it seems to me to be a minor modification is especially more logical for the comfort of the users.

Here is the place where it seems to me that the modification should take place :
That is example
```
function DocSettings:getSidecarDir(doc_path)
    if doc_path == nil or doc_path == '' then return '' end
    local file_without_suffix = doc_path:match("(.*)%.")
    if file_without_suffix then
        return file_without_suffix..".sdr"
    end
    return doc_path..".sdr"
end
```

Have see one workaround for fix my issue is to send configuration in history folder is better than create .sdr folder.

```
function DocSettings:getSidecarDir(doc_path)
    if doc_path == nil or doc_path == '' then return '' end
    local file_without_suffix = doc_path:match("(.*)%.")
    if file_without_suffix then
        return "."..file_without_suffix..".sdr"
    end
    return doc_path..".sdr"
end
```
**Please note, do not use this workaround, it is only for testing purposes !!!**

PS : I specify that I did the test and the hidden folders are well hidden by default. 

**Note 1 :** 
To avoid breaking compatibility, it would be preferable to be able to support both hidden folders and those that are not, via a regex for example.

**Note 2 :** 
Another very simple possibility too is not to simply create the .sdr files there where the ebook is located but in a dedicated folder.
ex: `koreader/config/Name.sdr`
But it should be expected that two ebooks have the same name. In this case you would have to do for ex: `koreader/config/Name_1.sdr `

Best Regard

**Comment by hius07** [comment]
The latest discussion about sdr is here: https://github.com/koreader/koreader/issues/8281.
You can find there more links showing the reasons of the current implementation.

**Comment by liberodark** [comment]
Sorry @hius07 but is not the same issue about sdr.
I dont want to remove .sdr after remove ebook.

**Comment by hius07** [comment]
> Another very simple possibility too is not to simply create the .sdr files there where the ebook is located but in a dedicated folder.

The links are in https://github.com/koreader/koreader/issues/8281#issuecomment-944340762.

**Comment by liberodark** [comment]
Thank you that great : https://github.com/koreader/koreader/issues/4831
Is trying to restore koreader/history.
I think is better solution.
Actually im making that with my workaround.

**Comment by poire-z** [comment]
It would be nice to have an option to use one or the other modes.
But it's quite a bit of work to get koreader/history/ to work in all situations, and just using this patch would have things breaking in some context (like when renaming a book, you would lose your settings). See https://github.com/koreader/koreader/issues/8281#issuecomment-931532274

**Comment by Neibiros** [comment]
Hi,
the problem is that `file_without_suffix` return path with filename e.g. `foo/bar.pdf` -> `foo/bar`. So your modification `return "."..file_without_suffix..".sdr"` return e.g. `.foo/bar.sdr`. It's is invalid path for creating folder, so Koreader use old History instead .sdr folder.

to create .sdr folder with prefix ".", I edit `function DocSettings:getSidecarDir(doc_path)` for my PocketBook HD 631 this way:
`--- Returns path to sidecar directory (`filename.sdr`).
--
-- Sidecar directory is the file without _last_ suffix.
-- @string doc_path path to the document (e.g., `/foo/bar.pdf`)
-- @treturn string path to the hidden sidecar directory (e.g., `/foo/.bar.sdr`)
function DocSettings:getSidecarDir(doc_path)
	if doc_path == nil or doc_path == '' then return '' end
	-- filename from doc_path with suffix (e.g., `bar.pdf`)
	local file_name = doc_path:match("^.+/(.+)$")
	-- filename from doc_path without suffix (e.g., `bar`)
	local file_name_without_suffix = doc_path:match("^.+/(.*)%.")
	-- filepath from doc_path (e.g., `/foo/`)
	local file_path = doc_path:match("^.+/")
	if file_name_without_suffix then
		return file_path.."."..file_name_without_suffix..".sdr"
	end
	return file_path.."."..file_name..".sdr"
end`
And now are .sdr folders hidden when I use default Pocketbook file manager. 

**Please note, do not use this workaround, it is only for testing purposes !!!**

**Comment by Jamesernator** [comment]
This has been particularly annoying on android especially for the Download directory as when browsing files in the regular file browser a lot of sdr folders hang around (and often orphaned as I've deleted their associated documents with Android's free space tool). 

It would make more sense if this book data were simply stored in whatever directory apps store their data.

---

This is probably a separate issue in itself, though the solution would be pretty similar, but the fact that renaming books and their associated metadata only works if you do it by KOReader is really inconvienent. Like I tend to organize things in my cloud storage and then just copy to my devices (both Android and Kobo), however if I've moved some files when I ultimately sync then read positions simply get lost if I've renamed the file (along with a bunch of orphaned `sdr` folders that no longer apply to anything).

This could be done just by saving the book data in a directory based on a hash of the file, so in this regard the solution is pretty similar to the other problem as both would require changing where `sdr` data is stored.

**Comment by Jamesernator** [comment]
> Or... you could just rename the folder along with the book? Problem solved ;o).

This is annoying which is the whole point here, if you read my comment, I note generally I would rename things in my cloud storage (onedrive) and just sync those folders onto my android and kobo devices. Like it would just be tedious to rename and/or move things around in cloud storage, sync to two devices and then update any changes on *both of the other devices*.

**Comment by NiLuJe** [comment]
Snark aside, something like

> This could be done just by saving the book data in a directory based on a hash of the file, so in this regard the solution is pretty similar to the other problem as both would require changing where `sdr` data is stored.

would be untenable. We'd have to compute a book's hash to even *display* basic metadata, and, which would be a massive performance hit, resulting in terrible UX.

It'd make some kind of sense if KOReader had a "metadata first" approach, but we don't, we have a *very* strong "file first" approach, so the filename is, and probably always will be, our main source of truth.

**Comment by NiLuJe** [comment]
TL;DR: The path of least resistance is indeed to update your tooling to account for SDR folders.

**Comment by NiLuJe** [comment]
And how do you associate a cache with a file/hash if you can't rely on the filename for that?

**Comment by NiLuJe** [comment]
> already does compute a hash

CRe does, once a book has been *actually* open'ed. It's unrelated to metadata handling. (Metadat handling is, also, technically unrelated to sidecar folders as far as the FM is concerned, but the source of truth for the db is still the absolute path of a file).

**Comment by Jamesernator** [comment]
Also this solution is more for the second problem I mentioned. For the first, which is just that all the sdr files are annoying when accessing `Download` in android's file browser, could be solved more simply just by changing .sdr to be under a KOReader config directory, i.e. rather than the all the hashing stuff just:

```
func getSdrDir(file) {
    /* Simply join the path with the koreader config dir so that sdr files aren't
       all floating around in Download */
    return joinPath(SDR_DIR_BASE, file)
}
```

Of the two problems I would say that the first annoys me more as there are regularly dozens to hundreds of sdr folders mixed in with my other things when browsing Download on android's file browser and it's just an absolute mess.

**Comment by NiLuJe** [comment]
See the patch I linked above for that ;).

**Comment by NiLuJe** [comment]
Keep in mind the K in KOReader still technically stands for "Kindle", which is the platform from which the sidecar folder implementation originated.

It's served us fairly well so far, despite a few similar discussions popping up from time to time.

As demonstrated by the patch above, moving to a centralized sidecar folder is relatively easily achievable, any other more involved core modifications are... probably not ;).

**Comment by KraXen72** [comment]
sorry, i read through this issue, however, i do not understand why sidecar directories couldn't start with a dot? that way, they exist for KOreader to interact with, but they're hidden for the user when browsing with a file manger, win-win.
thanks in advance!

**Comment by Frenzie** [comment]
The convention came from Kindle. I have no real opinion on it; it has both good and bad aspects.

**Comment by ell1e** [comment]
I just ran into this as well, the missing dot at the start of the directory. It's not that these sdr directories existing is a big problem, but without the dot they simply clutter any file manager listing of my book folder terribly.

**Comment by mergen3107** [comment]
My two cents, I don't really like the dot in the beginning. It makes one step further when you want to backup your book + its metadata- you'd need to enable viewing hidden files. 

On WinSCP I actually put a filter that makes all *.sdr folders colored blue, so I know immediately what is an actual folder and what's an sdr folder. But that's probably just me since I never use MTP after a JB and always stay on ssh. 

Also, in Linux you can sort all files and folders together, so book+metadata stay together in pairs in the list, very nice.

**Comment by mergen3107** [comment]
There are now two other ways to store sdr folders: hash and centralized folders, have a look at those, if interested:

http://koreader.rocks/user_guide/

Search for "Book metadata storage options"

**Comment by ell1e** [comment]
> Also, in Linux you can sort all files and folders together

That doesn't help when there are also non-sdr folders you still care about. Then it's still cluttered, just in another part of the listing.

In a central location has its own problems, e.g. when the book folder is on a network drive (which it is for me) then it won't be synchronized anymore between devices.

**Comment by ell1e** [comment]
Another use case I've come to ponder for using dot naming for the sdr folders is that the current approach risks confusing beginners. New non-techie users won't understand why suddenly strangely named folders keep cropping up in their books folder. It makes me reluctant to recommend KOReaders to newcomers, over something like Foliate that is easier to understand.

(I understand on an ereader device it wouldn't matter much, since those typically hide the folder structure anyway. But on e.g. a desktop PC, the situation is different.)

**Comment by mergen3107** [comment]
Then you should also not recommend Kindle at all, because this is what stock does too

**Comment by ell1e** [comment]
I don't, due to its imho worse DRM. I typically recommend buying DRM-free books and then using Foliate or MS Edge's built-in reader. Or if you can't help it and need to read DRM books on a regular machine, I typically point to Adobe Digital Editions.

**Comment by ell1e** [comment]
(Clarification note: also if we're talking about the physical kindle ebook readers, like I wrote above this problem is mainly existent on desktops. On physical ereaders the user typically isn't aware of the folder structure. But KOReader runs on regular desktops, too.)

**Comment by KraXen72** [comment]
fwiw, i switched to hashed due to this. i recommend others try out the other metadata storage options too, as was already mentioned here.
